tests/test_free_energy_v2.py

```diff
@@ -0,0 +1,173 @@
+    complex_box += np.eye(3)*0.1 # BFGS this later
+
+    # build the water system.
+    solvent_system, solvent_coords, solvent_box, _ = builders.build_water_system(4.0)
+    solvent_box += np.eye(3)*0.1 # BFGS this later
+
+    ff = Forcefield(deserialize_handlers(open('ff/params/smirnoff_1_1_0_ccc.py').read()))
+    ff_params = [
+        ff.hb_handle.params,
+        ff.ha_handle.params,
+        ff.pt_handle.params,
+        ff.it_handle.params,
+        ff.q_handle.params,
+        ff.lj_handle.params
+    ]
+
+    seed = 2021
+
+    lambda_schedule = np.linspace(0, 1.0, 4.0)
+    equil_steps = 1000
+    prod_steps = 1000
+
+    rfe = free_energy_v2.RelativeFreeEnergy(mol_a, mol_b, core, ff)
+
+    def vacuum_model(ff_params):
+
+        unbound_potentials, sys_params, masses, coords = rfe.prepare_vacuum_edge(ff_params)
+
+        x0 = coords
+        v0 = np.zeros_like(coords)
+        client = None
+        box = np.eye(3, dtype=np.float64)*100
+
+        integrator = LangevinIntegrator(
+            300.0,
+            1.5e-3,
+            1.0,
+            masses,
+            seed
+        )
+
+        model = estimator.FreeEnergyModel(
+            unbound_potentials,
+            client,
+            box,
+            x0,
+            v0,
+            integrator,
+            lambda_schedule,
+            equil_steps,
+            prod_steps
+        )
+
+        return estimator.deltaG(model, sys_params)
+
+    vg_fn = jax.value_and_grad(vacuum_model)
+    
+    vg_fn(ff_params) # dG and ff_params_grad
+
+    def binding_model(ff_params):
+
+        dGs = []
+
+        for host_system, host_coords, host_box in [
+            (complex_system, complex_coords, complex_box),
+            (solvent_system, solvent_coords, solvent_box)]:
+
+            # minimize the host to avoid clashes
+            host_coords = minimizer.minimize_host_4d(mol_a, host_system, host_coords, ff, host_box)
+
+            unbound_potentials, sys_params, masses, coords = rfe.prepare_host_edge(ff_params, host_system, host_coords)
+
+            x0 = coords
+            v0 = np.zeros_like(coords)
+            client = None
+            box = np.eye(3, dtype=np.float64)*100
+
+            integrator = LangevinIntegrator(
+                300.0,
+                1.5e-3,
+                1.0,
+                masses,
+                seed
+            )
+
+            model = estimator.FreeEnergyModel(
+                unbound_potentials,
+                client,
+                host_box,
+                x0,
+                v0,
+                integrator,
+                lambda_schedule,
+                equil_steps,
+                prod_steps
+            )
+
+            dG = estimator.deltaG(model, sys_params)
+            dGs.append(dG)
+
+        return dGs[0] - dGs[1]
+
+    # automatic chaining of vjps
+    vg_fn = jax.value_and_grad(binding_model)
```

Comment:
@maxentile I just wanted to briefly point you to this glorious line of code

Reply:
Nice! This could greatly simplify later steps, and make it easier to plug in different loss functions or likelihood functions down the road.